File: app/src/main/java/com/example/weizhenbin/wangebug/views/nestlayout/NestLayout.java
package com.example.weizhenbin.wangebug.views.nestlayout;

import android.content.Context;
import android.support.annotation.AttrRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.view.NestedScrollingParent2;
import android.support.v4.view.NestedScrollingParentHelper;
import android.support.v4.view.ViewCompat;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import com.example.weizhenbin.wangebug.R;
import com.example.weizhenbin.wangebug.views.TitleBar;

/**
 * Created by weizhenbin on 2018/9/21.
 */

public class NestLayout extends FrameLayout implements NestedScrollingParent2{

    IRollChange iRollChange;
    NestedScrollingParentHelper nestedScrollingParentHelper;

    TitleBar titleBar;

    int titleBarHeighet;
    public NestLayout(@NonNull Context context) {
        this(context,null);
    }

    public NestLayout(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs,0);
    }

    public NestLayout(@NonNull Context context, @Nullable AttributeSet attrs, @AttrRes int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        nestedScrollingParentHelper=new NestedScrollingParentHelper(this);
        titleBarHeighet=getResources().getDimensionPixelSize(R.dimen.title_bar_height);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();
        //找出所有符合条件的view
        //滑动的时候进行重新派遣指定事件
        titleBar= (TitleBar) getTargetView(this,TitleBar.class);

    }


    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        if (titleBar!=null) {
            setMeasuredDimension(getDefaultSize(0, widthMeasureSpec), getDefaultSize(0, heightMeasureSpec));
            int childWidthSize = getMeasuredWidth();
            int childHeightSize = getMeasuredHeight();
            widthMeasureSpec = MeasureSpec.makeMeasureSpec(childWidthSize, MeasureSpec.EXACTLY);
            /**按照比例改变高度值*/
            heightMeasureSpec = MeasureSpec.makeMeasureSpec(childHeightSize + titleBarHeighet, MeasureSpec.EXACTLY);
        }
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
    }

    //找到第一个符合条件的view
    private View getTargetView(ViewGroup viewGroup,Class aClass){
        if (viewGroup==null){
            return null;
        }
         int size=viewGroup.getChildCount();
        for (int i = 0; i < size; i++) {
            View v=viewGroup.getChildAt(i);
            if (v.getClass().equals(aClass)){
                return v;
            }else if(v instanceof ViewGroup){
              return this.getTargetView((ViewGroup) v,aClass);
            }
        }
        return null;
    }



    @Override
    public boolean onStartNestedScroll(@NonNull View child, @NonNull View target, int axes, int type) {
        return  (axes & ViewCompat.SCROLL_AXIS_VERTICAL) != 0;
    }

    @Override
    public void onNestedScrollAccepted(@NonNull View child, @NonNull View target, int axes, int type) {
         nestedScrollingParentHelper.onNestedScrollAccepted(child,target,axes,type);
    }

    @Override
    public void onStopNestedScroll(@NonNull View target, int type) {
        nestedScrollingParentHelper.onStopNestedScroll(target,type);
    }

    @Override
    public void onNestedScroll(@NonNull View target, int dxConsumed, int dyConsumed, int dxUnconsumed, int dyUnconsumed, int type) {
      //  Log.d("NestLayout", "onNestedScroll");
    }

    @Override
    public void onNestedPreScroll(@NonNull View target, int dx, int dy, @NonNull int[] consumed, int type) {
      //  Log.d("NestLayout", "onNestedPreScroll");
        if (titleBar!=null){
            //dy<0  向下拉
            if (dy<0&&getScrollY()>0){
                if (getScrollY()+dy<0){
                    scrollBy(0, -getScrollY());
                    consumed[1] = -getScrollY();
                }else {
                    scrollBy(0, dy);
                    consumed[1] = dy;
                }
            }
            //往上滑
            if (dy>0&&getScrollY()< titleBarHeighet){

                if (getScrollY()+dy>titleBarHeighet){
                    scrollBy(0, titleBarHeighet-getScrollY());
                    consumed[1] = titleBarHeighet-getScrollY();
                }else {
                    scrollBy(0, dy);
                    consumed[1] = dy;
                }
            }
            float alpha=1-(getScrollY()/titleBarHeighet);
            if (alpha<0.4){
                alpha=0;
            }
            titleBar.getLlTitleBar().setAlpha(alpha);
        }
        if (iRollChange!=null){
            iRollChange.onRollChange(dy);
        }
    }

    @Override
    public boolean onNestedFling(View target, float velocityX, float velocityY, boolean consumed) {
        return false;
    }

    @Override
    public boolean onNestedPreFling(View target, float velocityX, float velocityY) {
        return false;
    }


    public void setiRollChange(IRollChange iRollChange) {
        this.iRollChange = iRollChange;
    }
}

File: app/src/main/java/com/example/weizhenbin/wangebug/views/nestlayout/NestLayout.kt
package com.example.weizhenbin.wangebug.views.nestlayout

import android.content.Context
import android.support.annotation.AttrRes
import android.support.v4.view.NestedScrollingParent2
import android.support.v4.view.NestedScrollingParentHelper
import android.support.v4.view.ViewCompat
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout

import com.example.weizhenbin.wangebug.R
import com.example.weizhenbin.wangebug.views.TitleBar

/**
 * Created by weizhenbin on 2018/9/21.
 */

class NestLayout @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, @AttrRes defStyleAttr: Int = 0) : FrameLayout(context, attrs, defStyleAttr), NestedScrollingParent2 {

    private var iRollChange: IRollChange? = null
    private var nestedScrollingParentHelper: NestedScrollingParentHelper = NestedScrollingParentHelper(this)

    internal var titleBar: TitleBar? = null

    private var titleBarHeight: Int = 0

    init {
        titleBarHeight = resources.getDimensionPixelSize(R.dimen.title_bar_height)
    }

    override fun onFinishInflate() {
        super.onFinishInflate()
        //找出所有符合条件的view
        //滑动的时候进行重新派遣指定事件
        titleBar = getTargetView(this, TitleBar::class.java) as TitleBar?

    }


    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        var widthMeasureSpec = widthMeasureSpec
        var heightMeasureSpec = heightMeasureSpec
        if (titleBar != null) {
            setMeasuredDimension(View.getDefaultSize(0, widthMeasureSpec), View.getDefaultSize(0, heightMeasureSpec))
            val childWidthSize = measuredWidth
            val childHeightSize = measuredHeight
            widthMeasureSpec = View.MeasureSpec.makeMeasureSpec(childWidthSize, View.MeasureSpec.EXACTLY)
            heightMeasureSpec = View.MeasureSpec.makeMeasureSpec(childHeightSize + titleBarHeight, View.MeasureSpec.EXACTLY)
        }
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
    }

    //找到第一个符合条件的view
    private fun getTargetView(viewGroup: ViewGroup?, aClass: Class<*>): View? {
        if (viewGroup == null) {
            return null
        }
        val size = viewGroup.childCount
        for (i in 0 until size) {
            val v = viewGroup.getChildAt(i)
            if (v.javaClass == aClass) {
                return v
            } else if (v is ViewGroup) {
                return this.getTargetView(v, aClass)
            }
        }
        return null
    }


    override fun onStartNestedScroll(child: View, target: View, axes: Int, type: Int): Boolean {
        return axes and ViewCompat.SCROLL_AXIS_VERTICAL != 0
    }

    override fun onNestedScrollAccepted(child: View, target: View, axes: Int, type: Int) {
        nestedScrollingParentHelper.onNestedScrollAccepted(child, target, axes, type)
    }

    override fun onStopNestedScroll(target: View, type: Int) {
        nestedScrollingParentHelper.onStopNestedScroll(target, type)
    }

    override fun onNestedScroll(target: View, dxConsumed: Int, dyConsumed: Int, dxUnconsumed: Int, dyUnconsumed: Int, type: Int) {
        //  Log.d("NestLayout", "onNestedScroll");
    }

    override fun onNestedPreScroll(target: View, dx: Int, dy: Int, consumed: IntArray, type: Int) {
        //  Log.d("NestLayout", "onNestedPreScroll");
        if (titleBar != null) {
            //dy<0  向下拉
            if (dy < 0 && scrollY > 0) {
                if (scrollY + dy < 0) {
                    scrollBy(0, -scrollY)
                    consumed[1] = -scrollY
                } else {
                    scrollBy(0, dy)
                    consumed[1] = dy
                }
            }
            //往上滑
            if (dy > 0 && scrollY < titleBarHeight) {

                if (scrollY + dy > titleBarHeight) {
                    scrollBy(0, titleBarHeight - scrollY)
                    consumed[1] = titleBarHeight - scrollY
                } else {
                    scrollBy(0, dy)
                    consumed[1] = dy
                }
            }
            var alpha = (1 - scrollY / titleBarHeight).toFloat()
            if (alpha < 0.4) {
                alpha = 0f
            }
            titleBar!!.llTitleBar!!.alpha = alpha
        }
        if (iRollChange != null) {
            iRollChange!!.onRollChange(dy)
        }
    }

    override fun onNestedFling(target: View, velocityX: Float, velocityY: Float, consumed: Boolean): Boolean {
        return false
    }

    override fun onNestedPreFling(target: View, velocityX: Float, velocityY: Float): Boolean {
        return false
    }


    fun setiRollChange(iRollChange: IRollChange) {
        this.iRollChange = iRollChange
    }
}
